security/pkg/nodeagent/vm/nodeagent_v2.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vm
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/caclient"
+	caClientInterface "istio.io/istio/security/pkg/nodeagent/caclient/interface"
+	"istio.io/istio/security/pkg/nodeagent/plugin/providers/google/stsclient"
+	pkiutil "istio.io/istio/security/pkg/pki/util"
+	"istio.io/istio/security/pkg/platform"
+	"istio.io/istio/security/pkg/util"
+)
+
+// This file is similar to nodeagent.go, which implements NodeAgent interface.
+// However, this will implement NodeAgent using the new CA protocol which can be found here
+// https://github.com/istio/istio/blob/master/security/proto/istioca.proto
+
+type nodeAgentInternalV2 struct {
+	// Configuration specific to Node Agent
+	config   *Config
+	pc       platform.Client
+	caClient caClientInterface.Client
+	identity string
+	certUtil util.CertUtil
+}
+
+// Start starts the node Agent.
+func (na *nodeAgentInternalV2) Start() error {
+	if na.config == nil {
+		return fmt.Errorf("node Agent configuration is nil")
+	}
+
+	if !na.pc.IsProperPlatform() {
+		return fmt.Errorf("node Agent is not running on the right platform")
+	}
+
+	log.Infof("Node Agent V2 starts successfully.")
+
+	retries := 0
+	retrialInterval := na.config.CAClientConfig.CSRInitialRetrialInterval
+	identity, err := na.pc.GetServiceIdentity()
+	if err != nil {
+		return err
+	}
+	na.identity = identity
+	var success bool
+	for {
```

Comment:
Proposal:

- CitadelAgent.New(stop Channel), this way, we can stop the gorutine in a test.
-  Start() 

```bash
for {
  select 
    stopChannel <-
        stop
    timerChannel <- 
        req := sendRequest
        processRequest(timerChannel))
}

func processRequest(chan) {
  if success {
      // Use timer to wait for on the timer.
      returns
   }
   // Fails, should use backoff timer.
}
```

the specifics of the right way of using timer and channel you might need to lookup golang doc yourself.

The overall goal of is such the for loop becomes more compact and split the validate cert ttl/ fail or success wait timer in separate function.